src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+use candid::{CandidType, Deserialize, Principal};
+use ic_cdk::api::stable::stable64_size;
+use ic_cdk::api::time;
+use ic_cdk::{caller, trap};
+use ic_cdk_macros::{init, post_upgrade, query, update};
+use ic_stable_structures::memory_manager::{MemoryId, MemoryManager, VirtualMemory};
+use ic_stable_structures::{
+    cell::Cell as StableCell, log::Log, DefaultMemoryImpl, Memory as StableMemory,
+    RestrictedMemory, StableBTreeMap, Storable,
+};
+use internet_identity_interface::*;
+use metrics_encoder::MetricsEncoder;
+use serde_bytes::ByteBuf;
+use std::borrow::Cow;
+use std::cell::RefCell;
+
+type Memory = RestrictedMemory<DefaultMemoryImpl>;
+type StableLog = Log<VirtualMemory<Memory>, VirtualMemory<Memory>>;
+type ConfigCell = StableCell<ArchiveConfig, Memory>;
+type UserIndex = StableBTreeMap<VirtualMemory<Memory>, UserIndexKey, ()>;
+
+const GIB: u64 = 1 << 30;
+const WASM_PAGE_SIZE: u64 = 65536;
+/// The maximum number of Wasm pages that we allow to use for the stable storage.
+const MAX_WASM_PAGES: u64 = 8 * GIB / WASM_PAGE_SIZE;
+
+const LOG_INDEX_MEMORY_ID: MemoryId = MemoryId::new(0);
+const LOG_DATA_MEMORY_ID: MemoryId = MemoryId::new(1);
+const USER_INDEX_MEMORY_ID: MemoryId = MemoryId::new(2);
+
+/// Length of the user index key. Changing this value requires a stable memory migration.
+const USER_INDEX_KEY_LENGTH: usize = 24;
+
+thread_local! {
+    /// Static configuration of the archive that init() sets once.
+    static CONFIG: RefCell<ConfigCell> = RefCell::new(ConfigCell::init(
+        config_memory(),
+        ArchiveConfig::default(),
+    ).expect("failed to initialize stable cell"));
+
+    /// Static memory manager to manage the memory available for blocks.
+    static MEMORY_MANAGER: RefCell<MemoryManager<Memory>> = RefCell::new(MemoryManager::init(log_memory()));
+
+    /// Append-only list of encoded blocks stored in stable memory.
+    static LOG: RefCell<StableLog> = with_memory_manager(|memory_manager| {
+        RefCell::new(Log::init(memory_manager.get(LOG_INDEX_MEMORY_ID), memory_manager.get(LOG_DATA_MEMORY_ID)).expect("failed to initialize stable log"))
+    });
+
+    /// Index to efficiently filter entries by user number.
+    static USER_INDEX: RefCell<UserIndex> = with_memory_manager(|memory_manager| {
+        RefCell::new(StableBTreeMap::new(memory_manager.get(USER_INDEX_MEMORY_ID), USER_INDEX_KEY_LENGTH as u32, 0))
+    });
+}
+
+/// Creates a memory region for the configuration stable cell.
+fn config_memory() -> Memory {
+    RestrictedMemory::new(DefaultMemoryImpl::default(), 0..1)
```

Comment:
I'm a little uneasy about those bare indices for ranges, here and below. Are they related to the `..._MEMORY_ID`s that you define above?

Reply:
These bare indices just separate out the first memory page from the managed memory. I added documentation which explains this.